healthx/handler.go

```diff
@@ -78,18 +76,32 @@ func NewHandler(
 }
 
 type router interface {
-	GET(path string, handle httprouter.Handle)
+	Handler(method, path string, handler http.Handler)
 }
 
+type Middleware func(http.Handler) http.Handler
+
 // SetHealthRoutes registers this handler's routes for health checking.
-func (h *Handler) SetHealthRoutes(r router, shareErrors bool) {
-	r.GET(AliveCheckPath, h.Alive)
-	r.GET(ReadyCheckPath, h.Ready(shareErrors))
+func (h *Handler) SetHealthRoutes(r router, shareErrors bool, opts ...Middleware) {
```

Comment:
or are we talking about creating something similar to the proxy options setup you mentioned below?

Reply:
yes exactly, the options shouldn't just be a `http.Handlers`. Use the proxy as an example.

Reply:
Yeah, I have the base idea of implementing it, 
I wanted to confirm a requirement, do we wanna do multiple middlewares like  if someone was to write, he'd write it 
```go
WithMiddleware(
func1,
func2 
)
```

or do we wanna do single middleware and the developer decides if he wants to wrap them up

```go 
withMiddleware(func1)
```

based on that the code will change as below (you obviously know, but since it's also an assignment, just confirming) 

```go
type Options struct {
	Middleware func(http.Handler) http.Handler
}

// SetHealthRoutes registers this handler's routes for health checking.
func (h *Handler) SetHealthRoutes(r router, shareErrors bool, opts ...func(*Options)) {
	o := &Options{}

	for _, opt := range opts {
		opt(o)
	}

	r.Handler("GET", AliveCheckPath, o.Middleware(h.Alive()))
	r.Handler("GET", ReadyCheckPath, o.Middleware(h.Ready(shareErrors)))
}
```

vs 

```go
func (h *Handler) SetHealthRoutes(r router, shareErrors bool, opts ...func(*Options)) {
	o := &Options{}

	for _, opt := range opts {
		opt(o)
	}

	aliveHandler := h.Alive()
	readyHandler := h.Ready(shareErrors)

	for _, middleware := range o.Middleware {
		aliveHandler = middleware(aliveHandler)
		readyHandler = middleware(readyHandler)
	}

	r.Handler("GET", AliveCheckPath, aliveHandler)
	r.Handler("GET", ReadyCheckPath, readyHandler)
}
```

I prefer the second one cause then I can do the following 
```go
handler.SetHealthRoutes(router, true,
		WithMiddleware(
                   addMockHeaderMiddleware, 
                   testMiddleware(t, func(t *testing.T, rw http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "GET", r.Method)
		})),
	)
```

Reply:
@Benehiko Will push the needed code as soon as you confirm

Reply:
no i think you can keep it a single middleware since you can always wrap more middleware over the one you are passing down to the healthx package. The order of execution for middleware is sometimes really important, so I think it makes more sense to not implement something like that which could lead to future bugs because one middleware is executing after another instead of before.

Reply:
👍🏼 understood